engine/engine_launcher.go

```diff
@@ -241,35 +229,73 @@ func (el *Launcher) rollbackUpgrade(processLauncher rpc.LonghornProcessLauncherS
 	}
 
 	if el.backupEngine != nil {
-		el.lock.Unlock()
-		el.deleteEngine(processLauncher)
-		el.lock.Lock()
+		el.deleteEngine(processLauncher, processName)
 
-		el.currentEngine = el.backupEngine
-		el.backupEngine = nil
+		// need to wait for new engine process deletion before unset el.backupEngine.
+		isDeleted := el.waitForEngineProcessDeletion(processLauncher, processName)
+		if isDeleted {
+			el.lock.Lock()
+			el.backupEngine = nil
+			el.lock.Unlock()
+			logrus.Infof("engine launcher %v: successfully rolled back to backup engine", launcherName)
+		} else {
+			return fmt.Errorf("engine launcher %v: failed to rollback upgrade since new engine process cannot be deleted", launcherName)
+		}
 	}
 
-	logrus.Debugf("engine launcher %v: rollback completed", el.LauncherName)
-
 	return nil
 }
 
 func (el *Launcher) finalizeUpgrade(processLauncher rpc.LonghornProcessLauncherServiceServer) {
 	logrus.Debugf("engine launcher %v: finalize upgrade", el.LauncherName)
 
-	if err := el.deleteBackupEngine(processLauncher); err != nil {
+	el.lock.RLock()
+	launcherName := el.LauncherName
+	processName := el.backupEngine.EngineName
+	el.lock.RUnlock()
+
+	if _, err := el.deleteEngine(processLauncher, processName); err != nil {
 		logrus.Warnf("failed to delete backup engine process %v: %v", el.backupEngine.EngineName, err)
 	}
 
-	el.lock.Lock()
-	defer el.lock.Unlock()
+	// Need to wait for backup engine process deletion before unset el.backupEngine.
+	// Typically engine process deletion will cause frontend shutdown callback.
+	// But we don't want to shutdown frontend here since it's live upgrade.
+	// Hence frontend shutdown callback will check existence of el.backupEngine to avoid frontend down.
+	// Also need to block process since we cannot start another engine upgrade before completing this func.
+	isDeleted := el.waitForEngineProcessDeletion(processLauncher, processName)
+	if isDeleted {
+		el.lock.Lock()
+		el.binaryBackup = ""
+		el.backupEngine = nil
+		el.lock.Unlock()
+		logrus.Infof("engine launcher %v: successfully finalized backup engine", launcherName)
+	} else {
+		logrus.Errorf("engine launcher %v: failed to deleted backup engine process", el.LauncherName)
+	}
 
-	el.binaryBackup = ""
-	el.backupEngine = nil
+	return
+}
 
-	logrus.Debugf("engine launcher %v: finalize completed", el.LauncherName)
+func (el *Launcher) waitForEngineProcessDeletion(processLauncher rpc.LonghornProcessLauncherServiceServer, processName string) bool {
```

Comment:
Feel like we don't need to wait for failure. If it failed, print out an error and continue. Don't need to return false. If a process cannot be killed by SIGKILL we would have a problem, but we cannot do much about it.